### PR TITLE
Fix operator sync k8s deployment update

### PR DESCRIPTION
**Pull request by yittg**
# Description

Create a dapr k8s service for a deployment if it updates by adding a
dapr enabled annotation, or delete if the annotation removed.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #1831 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by yaron2** [comment]
Thanks for this @yittg

File: pkg/operator/handlers/dapr_handler.go [hunk 1]
```diff
@@ -186,6 +186,15 @@ func (h *DaprHandler) ObjectCreated(obj interface{}) {
 
 // ObjectUpdated handles Dapr crd updates
 func (h *DaprHandler) ObjectUpdated(old interface{}, new interface{}) {
+	oldAnnotated := h.isAnnotatedForDapr(old.(*appsv1.Deployment))
```
Comment (yaron2) [inline pkg/operator/handlers/dapr_handler.go:189]: The first if statement, lines 189 - 193, I'm not sure those are needed?

What this does is create a Dapr service if the old deployment doesn't have a Dapr annotation, which isn't a valid use case.

The second if statement is correct.

Comment (yaron2) [reply pkg/operator/handlers/dapr_handler.go:189]: The code should probably be this:

```go
annotated := h.isAnnotatedForDapr(new.(*appsv1.Deployment))
if !annotated {
	h.ObjectDeleted(new)
} else {
        h.ObjectCreated(new)
}
```

Comment (yittg) [reply pkg/operator/handlers/dapr_handler.go:189]: Actually the logics are equivalent. The current annotation is checked in the Create/Delete method.

It's ok to make it clearer though. 😃

Comment (yittg) [reply pkg/operator/handlers/dapr_handler.go:189]: And a bit more, we should pass the old deployment to Deleted method.

I think we should move the service add/delete logic out of the ObjectCreated/ObjectDeleted method, what do you think?

Comment (yaron2) [reply pkg/operator/handlers/dapr_handler.go:189]: > I think we should move the service add/delete logic out of the ObjectCreated/ObjectDeleted method, what do you think?

That's the only thing that the handler is in charge of, so I don't see a motivation to move that out. Where did you think to move it to?

Comment (yittg) [reply pkg/operator/handlers/dapr_handler.go:189]: I mean we should not call the Created/Deleted method in the Updated method directly.

Comment (yittg) [reply pkg/operator/handlers/dapr_handler.go]: let me show you the code :)

Comment (yaron2) [reply pkg/operator/handlers/dapr_handler.go]: Re: latest commits, this looks better, thanks.

**Comment by yaron2** [comment]
I'll re-run the tests to make sure it's not due to flakiness, but this may mean we need to adjust the way we setup our e2e test and understand what changed.

**Comment by yittg** [comment]
@yaron2 Where can i find the detail of the failed cases?

**Comment by yaron2** [comment]
https://github.com/dapr/dapr/runs/905393885?check_suite_focus=true

You can try and run those yourself, instructions are under the /tests dir.

**Comment by yittg** [comment]
@yaron2 i have fixed it, and tested locally. PTAL

**Comment by yaron2** [comment]
Thanks, the tests are passing now. What do you think caused the issues with having the deletion lock?

**Comment by yittg** [comment]
@yaron2 the lock has been moved into the newly added inner method, so the test cases failed due to the duplicated and not reentrant lock. Actually I should remove the original lock, and missed it,😀

**Pull request merged**